### DISCUSSION TITLE
Watch Folder

### DISCUSSION THREAD

**Discussion by Editpo** [RandomEngy/VidCoder#885] [Ideas]
Hi
Any chance of adding a Watch Folder for batch processing?

**Comment by RandomEngy** [comment]
Maybe someday. It would be a big feature that would require setting up a system service, and I'd need time to do it.

**Comment by Editpo** [comment]
Ok thanks. Ive been using HBBatchBeast but its a bit clunky

**Comment by jglazer63** [comment]
Given that there are command line options, you can do this with a little power scripting to watch a folder for files  and then add them to a queue.  This is precisely what I plan to do.

**Comment by RandomEngy** [comment]
Added a folder watcher in 8.4 Beta.

**Reply by jglazer63** [reply]
Where/How do you enable this?  I have 8.6 and cannot find the setting.

**Reply by RandomEngy** [reply]
@jglazer63 Windows -> Watcher. It is not supported in the Portable version, only in the Installer or zip.

**Comment by Editpo** [comment]
Excellent!
I've created a watch folder
ENABLED watch folder on the Windows icon.
I can see the Watch Folder window. It says Watcher - Starting... (is that meant to say Started after a bit?)
Put a file in the watch folder
Nothing happens!

Any clue?

**Comment by RandomEngy** [comment]
Looks like it is having trouble starting the VidCoderFileWatcher process. Can you check the %appdata%\VidCoder-Beta\Logs folder and see if there is a log named "Watcher"?

**Reply by Editpo** [reply]
I haven't got a LOGS folder

**Reply by Editpo** [reply]
Ah There is a logs folder in appdata/Roaming. Nothing called Watcher though

**Reply by Editpo** [reply]
The last entry in the logs folder is general and reads

# VidCoder 8.5 Beta
ERROR: Could not refresh folders in file watcher.
PipeMethodCalls.PipeInvokeFailedException: Can only invoke methods after connecting the pipe.
   at PipeMethodCalls.PipeInvokerHostExtensions.InvokeAsync[TRequesting](IPipeInvokerHost`1 invokerHost, Expression`1 expression, CancellationToken cancellationToken)
   at VidCoder.Services.WatcherProcessManager.RefreshFromWatchedFolders()

**Reply by RandomEngy** [reply]
Do you see a process named VidCoderFileWatcher running in Task Manager?

**Comment by Editpo** [comment]
No. Not seeing that

**Comment by Editpo** [comment]
FYI I have just installed into another PC. Still the same problem

**Reply by RandomEngy** [reply]
Sorry it took so long to get back to you, was out on vacation. I released v8.6 Beta, which has additional logging for the watcher process startup. Can you go to Global Options -> Advanced -> Logging and change Logging Verbosity to "Extended", then restart VidCoder and send me the entire "General" log?

**Reply by Editpo** [reply]
Hi
Updated to 8.6.
New g4eneral log has just 
# VidCoder 8.6 Beta
in it

**Reply by Editpo** [reply]
Ah. Forgot to turn on thw watcher. Now I get 
# VidCoder 8.6 Beta
# Waiting for watcher process to output a line.
# Line read from watcher process. Connecting now...
# Waiting for watcher process to output a line.
# Line read from watcher process. Connecting now...

**Reply by Editpo** [reply]
Eek! why is that so big :-)

**Reply by RandomEngy** [reply]
Hmm, so it launched the process and read a line from it. But you say you don't see any VidCoderFileWatcher process running. Can you take a screenshot of your Task Manager? Are you on the simple or advanced view? The simple view doesn't show background processes. If a VidCoderFileWatcher shows up briefly, then goes away, I would check for a crash in Event Viewer -> Windows Logs -> Application.

I'm also preparing some extra logging for the pipe connection part that apparently is failing.

**Reply by Editpo** [reply]
Hi
Here are screen shots
![Screenshot (2)](https://user-images.githubusercontent.com/87612977/189881395-d6baec80-9c5a-47db-afa6-65481e686d0d.png)
![Screenshot (3)](https://user-images.githubusercontent.com/87612977/189881398-952f9c41-d1a1-46d0-90fe-61e23fec557f.png)
![Screenshot (4)](https://user-images.githubusercontent.com/87612977/189881402-6ad85c5c-ba7d-4b7c-97c3-1754e8fd0837.png)
![Screenshot (5)](https://user-images.githubusercontent.com/87612977/189881404-4ac028e5-1d00-4457-9877-464647193cb8.png)

**Reply by Editpo** [reply]
There are two errors in event viewer
Faulting application name: VidCoderFileWatcher.exe, version: 8.6.0.0, time stamp: 0x62cf0990
Faulting module name: KERNELBASE.dll, version: 10.0.19041.1889, time stamp: 0xe9ede6d6
Exception code: 0xe0434352
Fault offset: 0x0000000000034fd9
Faulting process ID: 0x2c40
Faulting application start time: 0x01d8c75baddb0e64
Faulting application path: C:\Users\Lenovo M900\AppData\Local\VidCoder.Beta\app-8.6.0\VidCoderFileWatcher.exe
Faulting module path: C:\Windows\System32\KERNELBASE.dll
Report ID: 438bceec-3f96-4129-8485-220621f3da4b
Faulting package full name: 
Faulting package-relative application ID:

**Reply by Editpo** [reply]
2nd one

Application: VidCoderFileWatcher.exe
CoreCLR Version: 6.0.622.26707
.NET Version: 6.0.6
Description: The process was terminated due to an unhandled exception.
Exception Info: System.IO.DirectoryNotFoundException: Could not find a part of the path 'C:\Users\david\AppData\Roaming\VidCoder-Beta\Logs\2022-09-13 11.29.18 Watcher.txt'.
   at Microsoft.Win32.SafeHandles.SafeFileHandle.CreateFile(String fullPath, FileMode mode, FileAccess access, FileShare share, FileOptions options)
   at Microsoft.Win32.SafeHandles.SafeFileHandle.Open(String fullPath, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize)
   at System.IO.Strategies.OSFileStreamStrategy..ctor(String path, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize)
   at System.IO.Strategies.FileStreamHelpers.ChooseStrategyCore(String path, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize)
   at System.IO.Strategies.FileStreamHelpers.ChooseStrategy(FileStream fileStream, String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options, Int64 preallocationSize)
   at System.IO.StreamWriter.ValidateArgsAndOpenPath(String path, Boolean append, Encoding encoding, Int32 bufferSize)
   at System.IO.StreamWriter..ctor(String path)
   at VidCoderCommon.Services.SupportLogger..ctor(String type)
   at Program.<Main>$(String[] args)
   at Program.<Main>(String[] args)

**Reply by RandomEngy** [reply]
And _there_ is the smoking gun. Left in a line meant for debugging. Should be fixed in 8.7 Beta. Thanks for all your patience in helping me debug.

**Reply by Editpo** [reply]
Excellent! Its all working so far. Got 3 files in watch folder queue. Top one encoding
Glad you got there.

**Reply by Editpo** [reply]
Just noticed that if you put just one file in the watch folder it gets queued but doesn't start encoding. As soon as you add a second file it starts the queue.
Is this intentional?

**Reply by RandomEngy** [reply]
That's a bug that was introduced when I refactored away the scan dialog. It will be fixed in the next Beta release. Thanks for pointing it out.
